odoo/osv/expression.py

```diff
@@ -1031,12 +1036,10 @@ def push_result(leaf):
                     push(create_substitution_leaf(leaf, ('id', subop, (subquery, [ids2])), internal=True))
 
                 else:
-                    # determine ids1 = records with relations
-                    ids1 = select_distinct_from_where_not_null(cr, rel_id1, rel_table)
-
                     # rewrite condition to match records with/without relations
-                    op1 = 'in' if operator in NEGATIVE_TERM_OPERATORS else 'not in'
-                    push(create_substitution_leaf(leaf, ('id', op1, ids1), model))
+                    op1 = 'inselect' if operator in NEGATIVE_TERM_OPERATORS else 'not inselect'
+                    subquery = select_distinct_from_where_not_null_query(rel_id1, rel_table)
```

Comment:
You can remove the "DISTINCT" from the query in this case, as PG will figure out how to execute it efficiently (with a join or such).
```suggestion
                    subquery = 'SELECT "%s" FROM "%s" WHERE "%s" IS NOT NULL' % (rel_id1, rel_table, rel_id1)
```